typescript/web/src/connectors/apollo-client/cache-config.ts

```diff
@@ -0,0 +1,104 @@
+import {
+  InMemoryCacheConfig,
+  FieldFunctionOptions,
+  FieldReadFunction,
+  FieldMergeFunction,
+} from "@apollo/client";
+import type { QueryImagesArgs, Image } from "@labelflow/graphql-types";
+
+export type ImagesQueryCache = {
+  images: Image[];
+  totalCount: number | undefined;
+};
+
+const readQueryImages: FieldReadFunction<ImagesQueryCache, Image[]> = (
+  existing,
+  options: FieldFunctionOptions<QueryImagesArgs, Record<string, any>>
+) => {
+  // A read function should always return undefined if existing is
+  // undefined. Returning undefined signals that the field is
+  // missing from the cache, which instructs Apollo Client to
+  // fetch its value from your GraphQL server.
+  if (existing === undefined) {
+    return undefined;
+  }
+  const offset = options.args?.skip || 0;
+  const limit = options.args?.first || undefined;
+  const result =
+    offset > 0 || limit !== undefined
+      ? existing.images.slice(
+          offset,
+          limit !== undefined ? offset + limit : undefined
+        )
+      : existing.images;
+  if (
+    result.length === 0 ||
+    (result as (Image | undefined)[]).includes(undefined) ||
+    // Invalidate cache if returned result is smaller than the limit
+    // but does not match the expected total count of images in the
+    // dataset
+    (((limit && result.length < limit) || !limit) &&
+      (!existing.totalCount || existing.totalCount !== offset + result.length))
```

Comment:
That's what I call a big condition 😅 Can we split it into multiple well-named parts to reduce it's reading complexity?